js/shutterbug-support.ts

```diff
@@ -19,3 +21,37 @@ function beforeSnapshotHandler() {
     }
   });
 }
+
+export function takeSnapshot(selectorOrDomElement: string | HTMLElement) {
+  return new Promise((resolve, reject) => {
+    Shutterbug.snapshot({
+      selector: selectorOrDomElement,
+      done: (snapshotUrl: string) => {
+        resolve(snapshotUrl);
+      },
+      fail: (jqXHR: any, textStatus: string, errorThrown: any) => {
+        console.error("Shutterbug request failed", textStatus, errorThrown);
+        reject(errorThrown);
+      }
+    });
+  });
+}
+
+export function takePlanetViewSnapshot() {
+  return takeSnapshot(`#${PLANET_VIEW_CANVAS_ID}`);
+}
+
+export function takeCrossSectionSnapshot() {
+  const canvas = document.getElementById(CROSS_SECTION_CANVAS_ID);
+  let oldBgColor = "";
+  if (canvas) {
+    oldBgColor = canvas.style.backgroundColor;
+    canvas.style.backgroundColor = "black";
+  }
+  const promise = takeSnapshot(`#${CROSS_SECTION_CANVAS_ID}`);
```

Comment:
I know we don't want to wait until the full snapshot url is resolved but isn't there a risk that the background color is reset before the snapshot is taken?

Reply:
Fair question. I believe there's no risk, but mostly because I remember how Shutterbug JS lib works. :wink:
If there are no iframes, it'll clone and process HTML elements synchronously.
If there are iframes, it'll wait for their response, so even client-side processing will be async. And that would be the problem that you brought up.

We're dealing with canvas only, so the client-side processing will be fully synchronous.

There's no way to react to client-side processing only. If we wanted to be 100% sure, we'd need to wait for this promise to resolve. But that happens only after the server responds, so multiple seconds and it'll be visible for users.

I think it'd leave it this way, as even if we restore transparent background too soon, nothing bad will happen. Cross-section screenshot will just have a white background in the table. Black looks better, but the white one is fully functional. This juggling with the background is a minor aesthetic tweak.